### PR TITLE
Add CubicChunks support

### PR DESCRIPTION
**Pull request by NotStirred**
The commit history should be very clean, I didn't revert any changes so you should be able to go commit by commit to review

I'm going to do some more testing, and add javadoc to everything before this should be merged. This entire pr basically worked first try and that concerns me

Changes to existing classes:
- `Region` is now an interface, `MCRegion` is the vanilla implementation
- Create a `region` package for my own sanity
- All region objects are now created by the `World` implementation, previously several unrelated places
- Add a few min and max Y parameters on methods for chunk data

`ImposterCubicRegion` exists because CubicChunks regions are `256x256x256` blocks, while the vanilla ones are `1024*X*1024`, so internally the imposter takes the data from cubes for the requested height range and combines it into a chunk. This may be hundreds of regions vertically. It does that for each of the (currently) 4 region columns within its position

Also I may do some additional fixes for some oddities and bugs in the vanilla region stuff, let me know if that would be better in a separate PR after this one has been merged

### CODE REVIEW

**Comment by NotStirred** [comment]
Ready for review (probably)

**Comment by NotStirred** [comment]
Missing from this PR:
- parsing heightmap data (which is FAR more complex in cubic chunks)
- biomes, which are stored in two different places both 2d and 3d data, with 3d taking priority.
I will wait for chunky to support 3d biomes before adding this

**Review by aTom3333 — COMMENTED** [review]
Cubic chunk hype, this is so cool!

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private void loadBlockData0(Map<String, Tag> data, ChunkData chunkData, BlockPalette blockPalette, int minY, int maxY) {
```
[the remark this answers is not in the document]
Comment (NotStirred) [reply chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:150]: oh, forgot to remove it, yes it was a copy so I had a rough idea of what the vanilla implementation

File: chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+   * array of CubicRegion objects</p> */
+  private final Int2ReferenceOpenHashMap<CubicRegion112[]> internalRegions = new Int2ReferenceOpenHashMap<>();
+
+  private final Chunk[] chunks = new Chunk[CHUNKS_COUNT];
+  private final ChunkPosition position;
+  private final ChunkPosition min3drPosition;
+  private final World world;
+
+  /** Cubes don't have a timestamp, the hacky solution is to average the region timestamps.
+   * One timestamp per region column */
```
[the remark this answers is not in the document]
Comment (NotStirred) [reply chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java]: if one of the older regions change to a timestamp that's before the most recent one (such as could happen when merging worlds) it wouldn't be detect it. Though maybe that's not something we need to think about

File: chunky/src/java/se/llbit/chunky/world/CubicWorld.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+   */
+  protected CubicWorld(String levelName, File worldDirectory, int dimension, Set<PlayerEntityData> playerEntities,
+      boolean haveSpawnPos, long seed, long timestamp) {
+    super(levelName, worldDirectory, dimension, playerEntities, haveSpawnPos, seed, timestamp);
+  }
+
+  /**
+   * @return File object pointing to the region file directory
+   */
+  public synchronized File getRegionDirectory() {
```
[the remark this answers is not in the document]
Comment (NotStirred) [reply chunky/src/java/se/llbit/chunky/world/CubicWorld.java:37]: there are MANY methods that don't need to be synchronised, I copied from the existing implementation.
Perhaps it should be changed in a future PR? not sure

File: chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (aTom3333) [reply chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java]: I see, that's quite an edge case, good thinking. I guess I'm fine with your solution then

File: chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java [hunk 4]
```diff
@@ -433,23 +436,23 @@ public void parse() {
             }
 
             if (length < (long)sectorOffset * SECTOR_SIZE + Integer.BYTES) {
-              System.err.printf("Cube (%d, %d, %d) is outside of region file %s! Expected chunk data at offset %d but file length is %d%n",
-                cubicRegionToCube(regionX, localX), cubicRegionToCube(regionY, localY), cubicRegionToCube(regionZ, localZ), fileName, sectorOffset * SECTOR_SIZE_BYTES, length);
+              Log.warn(String.format("Cube (%d, %d, %d) is outside of region file %s! Expected chunk data at offset %d but file length is %d%n",
```
Comment (aTom3333) [inline chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java:439]: I think there is a Log.warnf function if you want (sorry for noting that earlier)

Comment (NotStirred) [reply chunky/src/java/se/llbit/chunky/world/region/ImposterCubicRegion.java:439]: oh... :facepalm:

**Comment by Dushess** [comment]
Good one, MCedit only left, but very different tool (editor)

**Comment by Agrrox** [comment]
> Good one, MCedit only left, but very different tool (editor)

[MCEdit](https://github.com/mcedit/mcedit2) is discontinued, use [Amulet Editor](https://github.com/Amulet-Team/Amulet-Map-Editor) instead

**Comment by NotStirred** [comment]
I would like to move the `RegionChangeWatcher` code into the world, so it can be implementation defined
thoughts @leMaik?

**Comment by leMaik** [comment]
@NotStirred Maybe `createRegionChangeWatcher` so that we don't have one giant class but rather a factory pattern? :thinking:

**Comment by NotStirred** [comment]
Sure, that sounds like java. Will implement it, then ready for review

File: chunky/src/java/se/llbit/chunky/ui/ChunkyFxController.java [hunk 5]
```diff
@@ -485,6 +490,12 @@ public synchronized void exportZip(File targetFile, ProgressTracker progress) {
     }));
     scale.valueProperty().addListener(new GroupedChangeListener<>(group,
         (observable, oldValue, newValue) -> mapView.setScale(newValue.intValue())));
+    yMin.valueProperty().addListener(new GroupedChangeListener<>(group, (observable, oldValue, newValue) -> {
+      if (!ignoreYMaxUpdate.get()) {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/ui/ChunkyFxController.java:494]:
```suggestion
      if (!ignoreYBoundsUpdate.get()) {
```

File: chunky/src/java/se/llbit/chunky/world/DeleteChunksJob.java [hunk 6]
```diff
@@ -17,6 +17,7 @@
 package se.llbit.chunky.world;
 
 import se.llbit.chunky.ui.ProgressTracker;
+import se.llbit.chunky.world.region.Region;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/DeleteChunksJob.java:20]:
```suggestion
```

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+public class ImposterCubicChunk extends Chunk {
+  private final CubicWorld world;
+
+  public ImposterCubicChunk(ChunkPosition pos, World world) {
+    super(pos, world);
+    assert world instanceof CubicWorld;
+    this.world = (CubicWorld) world;
+
+    assert world.getVersionId() <= 1343;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:29]: Put that version into some well-named constant? :)

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    version = "1.12";
+  }
+
+  private Map<Integer, Map<String, Tag>> getCubeTags(Set<String> request) {
+    Mutable<Integer> timestamp = new Mutable<>(dataTimestamp);
+    ImposterCubicRegion region = (ImposterCubicRegion) world.getRegion(position.getRegionPosition());
+    Map<Integer, Map<String, Tag>> cubeTagsInColumn = region.getCubeTagsInColumn(position, request, timestamp);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:36]: `Int2ObjectMap` or something?

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  public synchronized boolean loadChunk(ChunkData chunkData, int yMin, int yMax) {
+    if (!shouldReloadChunk()) {
+      return false;
+    }
+
+    Set<String> request = new HashSet<>();
+    request.add(Chunk.DATAVERSION);
+    request.add(Chunk.LEVEL_SECTIONS);
+    request.add(Chunk.LEVEL_BIOMES);
+    Map<Integer, Map<String, Tag>> data = getCubeTags(request);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:55]: [Int2ObjectMap](https://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/Int2ObjectMap.html)?

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    surfaceTimestamp = dataTimestamp;
+    loadSurfaceCubic(data, chunkData, yMin, yMax);
+    biomes = IconLayer.UNKNOWN;
+
+    biomesTimestamp = dataTimestamp;
+//    if (surface == IconLayer.MC_1_12) {
+//      biomes = IconLayer.MC_1_12;
+//    } else {
+//      loadBiomes(data, chunkData);
+//    }
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:70]:
```suggestion
//    TODO add biomes support once we have 3d biomes support
//    if (surface == IconLayer.MC_1_12) {
//      biomes = IconLayer.MC_1_12;
//    } else {
//      loadBiomes(data, chunkData);
//    }
```

File: chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    for (Map.Entry<Integer, Map<String, Tag>> entry : data.entrySet()) {
+      Integer yPos = entry.getKey();
+      Map<String, Tag> cubeData = entry.getValue();
+
+      Tag sections = cubeData.get(LEVEL_SECTIONS);
+      if (sections.isList()) {
+//        extractBiomeData(cubeData.get(LEVEL_BIOMES), chunkData);
+        if (version.equals("1.13") || version.equals("1.12")) {
+          loadBlockDataCubic(yPos, cubeData, chunkData, palette, yMin, yMax);
+          surface = new SurfaceLayer(world.currentDimension(), chunkData, palette, yMin, yMax);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/ImposterCubicChunk.java:93]: move this out of the loop?

File: chunky/src/java/se/llbit/chunky/world/World.java [hunk 12]
```diff
@@ -75,7 +81,7 @@
   /** Minimum level.dat data version of tall worlds (21w06a). */
   public static final int VERSION_21W06A = 2694;
 
-  private final Map<ChunkPosition, Region> regionMap = new HashMap<>();
+  protected final Map<ChunkPosition, Region> regionMap = new HashMap<>();
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/World.java:84]:
```suggestion
  protected final Long2ObjectMap<Region> regionMap = new Long2ObjectMap<>();
```

File: chunky/src/java/se/llbit/chunky/world/region/EmptyRegion.java [hunk 13]
```diff
@@ -32,18 +39,47 @@
...
+  }
+
+  @Override public Iterator<Chunk> iterator() {
+    return new Iterator<Chunk>() {
+      @Override public boolean hasNext() {
+        return false;
+      }
+
+      @Override public Chunk next() {
+        return EmptyRegionChunk.INSTANCE;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/region/EmptyRegion.java:79]:
```suggestion
        throw new NoSuchElementException();
```

File: chunky/src/java/se/llbit/chunky/world/region/MCRegion.java [hunk 14]
```diff
@@ -78,28 +85,22 @@ public Region(ChunkPosition pos, World world) {
...
-  public Chunk getChunk(ChunkPosition pos) {
-    return chunks[(pos.x & 31) + (pos.z & 31) * 32];
-  }
-
   /**
    * Set chunk at given position.
    */
-  public void setChunk(ChunkPosition pos, Chunk chunk) {
+  private void setChunk(ChunkPosition pos, Chunk chunk) {
     chunks[(pos.x & 31) + (pos.z & 31) * 32] = chunk;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/region/MCRegion.java:97]: We could use the `pos.getLong()` or something

**Review by leMaik — CHANGES_REQUESTED** [review]

**Pull request merged**